src/blas/KokkosBlas3_trmm.hpp

```diff
@@ -0,0 +1,179 @@
+///                   On entry, M-by-N matrix
+///                   On exit, overwritten with the solution
+template<class AViewType,
+         class BViewType>
+void
+trmm (const char side[],
+      const char uplo[],
+      const char trans[],
+      const char diag[],
+      typename BViewType::const_value_type& alpha,
+      const AViewType& A,
+      const BViewType& B)
+{
+
+  static_assert (Kokkos::Impl::is_view<AViewType>::value,
+                 "AViewType must be a Kokkos::View.");
+  static_assert (Kokkos::Impl::is_view<BViewType>::value,
+                 "BViewType must be a Kokkos::View.");
+  static_assert (static_cast<int> (AViewType::rank) == 2,
+                 "AViewType must have rank 2.");
+  static_assert (static_cast<int> (BViewType::rank) == 2,
+                 "BViewType must have rank 2.");
+
+  // Check validity of indicator argument
+  bool valid_side  = (side[0] == 'L' ) || (side[0] == 'l' )||
+                     (side[0] == 'R' ) || (side[0] == 'r' );
+  bool valid_uplo  = (uplo[0] == 'U' ) || (uplo[0] == 'u' )||
+                     (uplo[0] == 'L' ) || (uplo[0] == 'l' );
+  bool valid_trans = (trans[0] == 'N') || (trans[0] == 'n')||
+                     (trans[0] == 'T') || (trans[0] == 't')||
+                     (trans[0] == 'C') || (trans[0] == 'c');
+  bool valid_diag  = (diag[0] == 'U' ) || (diag[0] == 'u' )||
+                     (diag[0] == 'N' ) || (diag[0] == 'n' );
+  if(!valid_side) {
+    std::ostringstream os;
+    os << "KokkosBlas::trmm: side = '" << side[0] << "'. " <<
+      "Valid values include 'L' or 'l' (A is on the left of X), "
+      "'R' or 'r' (A is on the right of X).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  if(!valid_uplo) {
+    std::ostringstream os;
+    os << "KokkosBlas::trmm: uplo = '" << uplo[0] << "'. " <<
+      "Valid values include 'U' or 'u' (A is upper triangular), "
+      "'L' or 'l' (A is lower triangular).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  if(!valid_trans) {
+    std::ostringstream os;
+    os << "KokkosBlas::trmm: trans = '" << trans[0] << "'. " <<
+      "Valid values include 'N' or 'n' (No transpose), 'T' or 't' (Transpose), "
+      "and 'C' or 'c' (Conjugate transpose).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  if(!valid_diag) {
+    std::ostringstream os;
+    os << "KokkosBlas::trmm: diag = '" << diag[0] << "'. " <<
+      "Valid values include 'U' or 'u' (the diagonal of A is assumed to be unit), "
+     "'N' or 'n' (the diagonal of A is assumed to be non-unit).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+
+  bool is_A_lower_triangle = (side[0] == 'L' || side[0] == 'l');
+
+  int64_t A_m = A.extent(0);
+  int64_t A_n = A.extent(1);
+  int64_t B_m = B.extent(0);
+  int64_t B_n = B.extent(1);
+
+  // Return if degenerated matrices are provided
+  if((A_m == 0) || (A_n == 0) || (B_m == 0) || (B_n == 0))
```

Comment:
You don't need the inner parentheses, and some compilers warn about them.  `||` has lower precedence than `==`.